### PR TITLE
Setup Detekt

### PR DESCRIPTION
**Pull request by racka98**
## 📑 What does this PR do?

This PR adds support for Detekt as discussed on issue #11

The rule add here are the default ones with some tweaks for Jetpack Compose.
Such function naming, complexity, compose annotations and [Twitter Comose Rules](https://github.com/twitter/compose-rules)

# ✅ Checklist

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [ ] All the tests have passed

## 🧪 How can this PR been tested?

- Run `./gradlew detekt` to run Detekt for the entire project
- Run `./gradlew module:detekt` to run Detekt on a specific module

## 🧾 Tasks Remaining: (List of tasks remaining to be implemented)

- Running the check for the whole project and fixing all the issues highlighted by Detekt. This should be done on a new PR after this has been merged.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -53,8 +53,9 @@ Atomic commits will make it easier to track down regressions. Also, it enables t
 1. Fork it (https://github.com/AndroidDev-social/MastodonCompose/fork)
 2. Create a new feature branch (`git checkout -b feature/fooBar`)
 3. Commit your changes (`git commit -am 'Add some fooBar'`)
-4. Push to the branch (`git push origin feature/fooBar`)
-5. Create a new Pull Request
+4. Run `./gradlew detekt` and fix any code issues before pushing
```
Comment (thebino) [inline CONTRIBUTING.md:56]: I would run `./gradlew check` to run all other tests and checks as well.

File: README.md [hunk 2]
```diff
@@ -44,6 +44,15 @@ Alternatively, there can be contributions made that aren't code related.
 - General question, gotchas and other discussions [here](https://github.com/AndroidDev-social/MastodonCompose/discussions/categories/general)
 - Create new Issues/Features for post MVP. (e.g. Material You support, animations, V2 functionality, migrations, etc)
 
+
+### Detekt
```
Comment (thebino) [inline README.md:48]: I would omit developer infos in the readme.
It is fine to have it in the contribution file only

File: CONTRIBUTING.md (hunk 1 above)
Comment (racka98) [reply CONTRIBUTING.md:56]: Good idea. Though Detekt hasn't been enabled to work with the check task yet (until everything has been properly formatted we can enable it)

File: README.md (hunk 2 above)
Comment (racka98) [reply README.md:48]: So I should remove that whole block?

Comment (thebino) [reply README.md:48]: yes please. If necessary move it to the contribution.md

The Readme is also for non-dev people. They don't care about these technical details

File: CONTRIBUTING.md (hunk 1 above)
Comment (thebino) [reply CONTRIBUTING.md:56]: Okay, then we can keep it for now and change it later when it is fixed and enabled.

Comment (racka98) [reply CONTRIBUTING.md:56]: Should be changed on the next PR that fixes all the detekt issues

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -155,3 +156,28 @@ To build and run the android app (emulator or real android device is needed)
...
+
+1. Run for whole project: 
+```shell
+$ ./gradlew detekt
+```
+
+2. Run for specific module:
+```shell
+$ ./gradlew module:detekt
+```
```
Comment (racka98) [inline CONTRIBUTING.md:183]: @thebino the `Verifications` details have all been moved here. Hope that solves it

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -155,3 +156,28 @@ To build and run the android app (emulator or real android device is needed)
...
+### 🔬 Verifications
+
+To verify important checks, run the following commands on your local machine
+
+#### General Check
+
+Run all configured checks for the project
+```shell
+$ ./gradlew check
+```
```
Comment (crocsandcoffee) [inline CONTRIBUTING.md:169]: what all does this run? just unit tests?

File: build.gradle.kts [hunk 5]
```diff
@@ -50,3 +78,79 @@ subprojects {
...
+            io.gitlab.arturbosch.detekt.extensions.DetektExtension.DEFAULT_TEST_SRC_DIR_JAVA,
+            io.gitlab.arturbosch.detekt.extensions.DetektExtension.DEFAULT_SRC_DIR_KOTLIN,
+            io.gitlab.arturbosch.detekt.extensions.DetektExtension.DEFAULT_TEST_SRC_DIR_KOTLIN,
+            // Kotlin Multiplatform
+            "src/commonMain/kotlin",
+            "src/androidMain/kotlin",
+            "src/iosMain/kotlin",
+            "src/jvmMain/kotlin",
+            "src/desktopMain/kotlin",
+            "src/jsMain/kotlin",
```
Comment (crocsandcoffee) [inline build.gradle.kts:107]: should we also include test?

File: build.gradle.kts [hunk 6]
```diff
@@ -50,3 +78,79 @@ subprojects {
...
+        exclude("**/build/**", "**/generated/**", "**/resources/**")
+        basePath = rootProject.projectDir.absolutePath
+    }
+}
+
+/**
+ * By default Detekt runs when the Gradle :check task is triggered
+ * It should be disabled for now until everything has been correctly setup (including CI)
+ */
+fun Project.filterDetektFromCheckTask() {
```
Comment (crocsandcoffee) [inline build.gradle.kts:148]: make private

File: build.gradle.kts [hunk 7]
```diff
@@ -50,3 +78,79 @@ subprojects {
         }
     }
 }
+
+/**
+ * To run detekt simply:
+ * 1. ./gradlew module:detekt for each module
+ * 2. ./gradlew detekt for whole project
+ */
+fun Project.coreDetektSetup() {
```
Comment (crocsandcoffee) [inline build.gradle.kts:87]: make private

File: build.gradle.kts [hunk 8]
```diff
@@ -50,3 +78,79 @@ subprojects {
...
+/**
+ * By default Detekt runs when the Gradle :check task is triggered
+ * It should be disabled for now until everything has been correctly setup (including CI)
+ */
+fun Project.filterDetektFromCheckTask() {
+    if (tasks.names.contains("check")) {
+        tasks.named("check").configure {
+            this.setDependsOn(this.dependsOn.filterNot {
+                it is TaskProvider<*> && it.name == "detekt"
+            })
```
Comment (crocsandcoffee) [inline build.gradle.kts:153]: does this filter out both detekt and ktlint?

**Review by crocsandcoffee — APPROVED** [review]
LGTM - just a couple sanity questions for you

File: CONTRIBUTING.md (hunk 4 above)
Comment (racka98) [reply CONTRIBUTING.md:169]: The check task should run all tests and detekt (once allowed)

File: build.gradle.kts (hunk 5 above)
Comment (racka98) [reply build.gradle.kts:107]: Yes imo. They contain code too

File: build.gradle.kts (hunk 6 above)
Comment (racka98) [reply build.gradle.kts:148]: There's no `private` modifier in build scripts

File: build.gradle.kts (hunk 8 above)
Comment (racka98) [reply build.gradle.kts:153]: Only detekt. That will be removed anyways after all the code has been properly formatted and sanitized

File: build.gradle.kts (hunk 6 above)
Comment (crocsandcoffee) [reply build.gradle.kts:148]: TIL! 😄

File: build.gradle.kts (hunk 5 above)
Comment (crocsandcoffee) [reply build.gradle.kts:107]: Okay, please add them, they follow the same nomenclature. (commonTest, etc)

Comment (racka98) [reply build.gradle.kts]: Done

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -55,8 +55,9 @@ Atomic commits will make it easier to track down regressions. Also, it enables t
 2. Install pre-commit hooks (`git config core.hooksPath .githooks`)
 3. Create a new feature branch (`git checkout -b feature/fooBar`)
 4. Commit your changes (`git commit -am 'Add some fooBar'`)
-5. Push to the branch (`git push origin feature/fooBar`)
-6. Create a new Pull Request
+5. Run `./gradlew detekt` and fix any code issues before pushing
```
Comment (thebino) [inline CONTRIBUTING.md:58]: not necessary since it is covered by git pre-commit

File: CONTRIBUTING.md (hunk 3 above)
Comment (thebino) [reply CONTRIBUTING.md:189]: lgtm

File: CONTRIBUTING.md (hunk 9 above)
Comment (racka98) [reply CONTRIBUTING.md]: Done

**Pull request closed without merging**